pkg/maps/cidrmap/cidrmap.go

```diff
@@ -67,21 +68,74 @@ func (cm *CIDRMap) cidrKeyInit(cidr net.IPNet) (key cidrKey) {
 	return
 }
 
-// AllowCIDR inserts an entry to 'cm' with key 'cidr'. Value is currently not
+func (cm *CIDRMap) keyCidrInit(key cidrKey) (cidr net.IPNet) {
+	cidr.Mask = net.CIDRMask(int(key.Prefixlen), int(cm.AddrSize*8))
+	cidr.IP = make(net.IP, cm.AddrSize)
+	copy(cidr.IP[len(cidr.IP)-cm.AddrSize:len(cidr.IP)], key.Net[:])
+	return
+}
+
+// InsertCIDR inserts an entry to 'cm' with key 'cidr'. Value is currently not
 // used.
-func (cm *CIDRMap) AllowCIDR(cidr net.IPNet) error {
+func (cm *CIDRMap) InsertCIDR(cidr net.IPNet) error {
 	key := cm.cidrKeyInit(cidr)
 	entry := [LPM_MAP_VALUE_SIZE]byte{}
+	if cm.Prefixlen != 0 && cm.Prefixlen != key.Prefixlen {
+		return fmt.Errorf("Unable to update element with different prefixlen than map!")
+	}
 	return bpf.UpdateElement(cm.Fd, unsafe.Pointer(&key), unsafe.Pointer(&entry), 0)
 }
 
+// DeleteCIDR deletes an entry from 'cm' with key 'cidr'.
+func (cm *CIDRMap) DeleteCIDR(cidr net.IPNet) error {
+	key := cm.cidrKeyInit(cidr)
+	if cm.Prefixlen != 0 && cm.Prefixlen != key.Prefixlen {
+		return fmt.Errorf("Unable to delete element with different prefixlen than map!")
+	}
+	return bpf.DeleteElement(cm.Fd, unsafe.Pointer(&key))
+}
+
 // CIDRExists returns true if 'cidr' exists in map 'cm'
 func (cm *CIDRMap) CIDRExists(cidr net.IPNet) bool {
 	key := cm.cidrKeyInit(cidr)
 	var entry [LPM_MAP_VALUE_SIZE]byte
 	return bpf.LookupElement(cm.Fd, unsafe.Pointer(&key), unsafe.Pointer(&entry)) == nil
 }
 
+// CIDRNext returns next CIDR entry in map 'cm'
+func (cm *CIDRMap) CIDRNext(cidr *net.IPNet) *net.IPNet {
+	var key, keyNext cidrKey
+	if cidr != nil {
+		key = cm.cidrKeyInit(*cidr)
+	}
+	err := bpf.GetNextKey(cm.Fd, unsafe.Pointer(&key), unsafe.Pointer(&keyNext))
+	if err != nil {
+		return nil
+	}
+	out := cm.keyCidrInit(keyNext)
+	return &out
+}
+
+// CIDRDump walks map 'cm' and dumps all CIDR entries
+func (cm *CIDRMap) CIDRDump(to []string) []string {
```

Comment:
You can create the return slice inside the function, no need to pass it as an argument.

```
func (cm *CIDRMap) CIDRDump() []string {
to := []string{}
```

Reply:
This is passed by value and not reference in golang, right? So it would otherwise have to copy the slice around? In that case, yes I'll fix it.

Reply:
Okay, it's just the slice header, but not the backing array, so should be fine as is.